File: app/src/classes/Person.js
/**
 * Person Class (Abstract)
 * Demonstrates: Abstraction and Encapsulation
 * Base class for all person entities in the system
 */

export class Person {
  // Private attributes - Encapsulation
  #id;
  #name;

  constructor(id, name) {
    if (new.target === Person) {
      throw new TypeError("Cannot instantiate abstract class Person");
    }
    this.#id = id;
    this.#name = name;
  }

  // Getters - Encapsulation
  getId() {
    return this.#id;
  }

  getName() {
    return this.#name;
  }

  // Setters - Encapsulation
  setId(id) {
    this.#id = id;
  }

  setName(name) {
    this.#name = name;
  }

  // Abstract method - Polymorphism
  // Subclasses must implement this method
  display() {
    throw new Error("display() method must be implemented in subclass");
  }

  // Concrete method in abstract class
  printBasicInfo() {
    return `ID: ${this.#id} | Name: ${this.#name}`;
  }

  // Additional utility method
  getInfo() {
    return {
      id: this.#id,
      name: this.#name
    };
  }
}

File: java-app/src/Person.java
/**
 * Abstract Person Class
 * Demonstrates: Abstraction and Encapsulation
 * Base class for all person entities in the system
 */
public abstract class Person {
    // Encapsulation: Private attributes
    private String id;
    private String name;

    // Constructor
    public Person(String id, String name) {
        this.id = id;
        this.name = name;
    }

    // Getters - Encapsulation
    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    // Setters - Encapsulation
    public void setId(String id) {
        this.id = id;
    }

    public void setName(String name) {
        this.name = name;
    }

    // Abstract method - Polymorphism
    // Subclasses must implement this method
    public abstract void display();

    // Concrete method in abstract class
    public void printBasicInfo() {
        System.out.println("ID: " + id + " | Name: " + name);
    }
}
